driver/controller.go

```diff
@@ -212,30 +199,71 @@ func (d *Driver) ControllerPublishVolume(ctx context.Context, req *csi.Controlle
 
+	// check if volume exist before trying to attach it
+	vol, err := d.cloudscaleClient.Volumes.Get(ctx, req.VolumeId)
+	if err != nil {
+		errorResponse, ok := err.(*cloudscale.ErrorResponse)
+		if ok && errorResponse.StatusCode == http.StatusNotFound {
+			return nil, status.Errorf(codes.NotFound, "volume %q not found", req.VolumeId)
+		}
+		return nil, err
+	}
+
+	// check if the server exist before trying to attach the volume to it
+	_, err = d.cloudscaleClient.Servers.Get(ctx, req.NodeId)
+	if err != nil {
+		errorResponse, ok := err.(*cloudscale.ErrorResponse)
+		if ok && errorResponse.StatusCode == http.StatusNotFound {
+			return nil, status.Errorf(codes.NotFound, "server %q not found", req.NodeId)
+		}
+		return nil, err
+	}
+
+	if vol.ServerUUIDs != nil {
+		// check if the volume is already attached to the server
+		for _, serverId := range *vol.ServerUUIDs {
+			if serverId == req.NodeId {
+				ll.Info("volume is already attached")
+				return &csi.ControllerPublishVolumeResponse{
+					PublishContext: map[string]string{
+						PublishInfoVolumeName: vol.Name,
+					},
+				}, nil
+			}
+		}
+		// volume is attached to a different server
+		if len(*vol.ServerUUIDs) > 0 {
+			serverId := (*vol.ServerUUIDs)[0]
+			return nil, status.Errorf(codes.FailedPrecondition,
+				"volume %q is attached to the wrong server (%q), detach the volume to fix it", req.VolumeId, serverId)
+		}
+	}
+
 	attachRequest := &cloudscale.Volume{
 		ServerUUIDs: &[]string{req.NodeId},
 	}
-	err := d.cloudscaleClient.Volumes.Update(ctx, req.VolumeId, attachRequest)
+	err = d.cloudscaleClient.Volumes.Update(ctx, req.VolumeId, attachRequest)
 	if err != nil {
-		/*
-			// don't do anything if attached
-			if resp != nil && resp.StatusCode == http.StatusUnprocessableEntity {
-				if strings.Contains(err.Error(), "Droplet already has a pending event") {
-					ll.WithFields(logrus.Fields{
-						"error": err,
-						"resp":  resp,
-					}).Warn("droplet is not able to detach the volume")
-					// sending an abort makes sure the csi-attacher retries with the next backoff tick
-					return nil, status.Errorf(codes.Aborted, "volume %q couldn't be attached. droplet %s is in process of another action",
-						req.VolumeId, req.NodeId)
-				}
-			}
-		*/
-		return nil, reraiseNotFound(err)
+		errorResponse, ok := err.(*cloudscale.ErrorResponse)
+		if ok && errorResponse.StatusCode == http.StatusNotFound {
+			ll.WithFields(logrus.Fields{
+				"error":         err,
+				"errorResponse": errorResponse,
+			}).Warnf("server %q or volume %q not found", req.NodeId, req.VolumeId)
+			return nil, status.Errorf(codes.NotFound, "server %q or volume %q not found", req.NodeId, req.VolumeId)
+		}
+		ll.WithFields(logrus.Fields{
+			"error": err,
+		}).Warnf("volume %q couldn't be attached to server %q", req.VolumeId, req.NodeId)
+		return nil, status.Errorf(codes.Aborted, "volume %q couldn't be attached to server %q", req.VolumeId, req.NodeId)
```

Comment:
Why was the old way not good enough? I personally like `reraiseNotFound()`

Reply:
My initial goal was to keep the call to `reraiseNotFound` instead of explicitly handling not found like digitalocean does.

However,  `reraiseNotFound` doesn't behave like the rest of the controller's code with regards to logging, so I thought it would make more sense to remove the call to `reraiseNotFound`. 

Another reason is the return code where `reraiseNotFound` doesn't set `codes.Aborted` if it's an error different from a 404. I'm not entirely sure what the semantics are if the CSI components from the CO receive a generic error.

Reply:
My initial goal was to keep the call to `reraiseNotFound` instead of explicitly handling not found like digitalocean does.

However, `reraiseNotFound` doesn't behave like the rest of the controller's code with regards to logging, so I thought it would make more sense to remove the call to `reraiseNotFound`.

Another reason is the return code where `reraiseNotFound` doesn't set `codes.Aborted` if it's an error different from a 404. I'm not entirely sure what the semantics are if the CSI components from the CO receive a generic error.